trajectory_executor/src/trajectory_executor/trajectory_executor.cpp

```diff
@@ -0,0 +1,188 @@
+                std::map<std::string, ros::Publisher>::iterator it = _traj_publisher_map.find(control_plugin);
+                if (it != _traj_publisher_map.end()) {
+                    ROS_DEBUG("Found match for control plugin %s at point %d in current trajectory!",
+                        control_plugin.c_str(),
+                        _timesteps_since_last_traj);
+                    it->second.publish(*_cur_traj);
+                } else {
+                    std::ostringstream description_builder;
+                    description_builder << "No match found for control plugin " 
+                        << control_plugin << " at point " 
+                        << _timesteps_since_last_traj << " in current trajectory!";
+                    shutdown(description_builder.str());
+                }
+                _timesteps_since_last_traj++;
+            } else {
+                shutdown("Ran out of trajectory data to consume!");
+            }
+        } else {
+            ROS_DEBUG("Awaiting initial trajectory publication...");
+        }
+        ROS_DEBUG("TrajectoryExecutor tick completed succesfully!");
+    }
+
+    void TrajectoryExecutor::shutdown(std::string reason) {
+        ROS_ERROR_STREAM("FATAL ERROR: " << reason);
+        cav_msgs::SystemAlert system_alert;
+        system_alert.type = system_alert.FATAL;
+        system_alert.description = reason;
+        ROS_FATAL("Publishing FATAL error on system alert topic!");
+        _default_nh->publishSystemAlert(system_alert);
+        ros::shutdown();
+
+        exit(1);
+    }
+
+    void TrajectoryExecutor::run()
+    {
+        ROS_DEBUG("Starting operations for TrajectoryExecutor component...");
+        _timer = _timer_nh->createTimer(
+            ros::Duration(ros::Rate(this->_min_traj_publish_tickrate_hz)),
+            &TrajectoryExecutor::onTrajEmitTick, 
+            this);
+
+        ros::AsyncSpinner timer_spinner(1, &_timer_callbacks);
+        timer_spinner.start();
+        ROS_DEBUG("Timer thread started!");
+
+        ros::AsyncSpinner msg_spinner(1, &_msg_callbacks);
+        msg_spinner.start();
+
+        ROS_DEBUG("TrajectoryExecutor component started succesfully! Starting to spin.");
+
+        ros::Rate r(_default_spin_rate);
+        while (ros::ok())
+        {
+            ros::spinOnce();
+            r.sleep();
+        }
+
+        timer_spinner.stop();
+        msg_spinner.stop();
+
+        ros::shutdown();
+    }
+
+    bool TrajectoryExecutor::init()
+    {
+        ROS_DEBUG("Initializing TrajectoryExecutor node...");
+    
+        _default_nh = std::unique_ptr<ros::CARMANodeHandle>(new ros::CARMANodeHandle);
+        _timer_nh = std::unique_ptr<ros::CARMANodeHandle>(new ros::CARMANodeHandle);
+        _timer_nh->setCallbackQueue(&_timer_callbacks);
+        _msg_nh = std::unique_ptr<ros::CARMANodeHandle>(new ros::CARMANodeHandle);
+        _msg_nh->setCallbackQueue(&_msg_callbacks);
```

Comment:
Usually, we are using one general NodeHandle and a private NodeHandle and using their global default callback queue. What is the reason to have three here?

Reply:
The idea is to allow the timer thread (which uses a TimerEvent in the callback queue) to run at whatever rate it needs to without depending on the rate of the message processing queue and to allow the message queue to spin as fast as it receives messages (through use of an async spinner). I have made the `_default_nh` private so that we can read params from it as you mentioned, though.